<a href="https://colab.research.google.com/github/IsabellaBonnet/analitica/blob/main/1_Reto_Isabella_Bonnet.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Modelo de Clasificación Naive Bayes**

0. Se procede con la carga de las librerias de trabajo

In [ ]:
# Importar las librerías necesarias
import pandas as pd #Comunicación con Excel
import numpy as np #Funciones Matemáticas
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB #Función Gaussiana
from sklearn.metrics import confusion_matrix #Métricas de medida
from sklearn.metrics import classification_report, accuracy_score

1. se cargan los datos

In [ ]:
nxl="/content/1. BD1_Longitudinal Survey_Int.xlsx"
XDB=pd.read_excel(nxl,sheet_name=0).dropna() #Traigo los datos
 #Eliminar datos faltantes, celdas vacias

#Seleccionamos las variables de interes
XDB=XDB[['Age','Mother_Edu','Father_Edu','Height','Weight','Income','Self_Esteem','FamilySize','Urban']]
XDB

#Datos de Entrada
XD=XDB[['Age','Mother_Edu','Father_Edu','Height','Weight','Income','Self_Esteem','FamilySize']]
#Datos de salida - Valor deseado
yd=np.array(XDB['Urban'] )
print (XDB)

       Age  Mother_Edu  Father_Edu  Height  Weight   Income  Self_Esteem  \
1       20         5.0         8.0    62.0   120.0      0.0         16.0   
5       19        12.0        12.0    64.0   200.0  40000.0         27.0   
6       15        12.0        12.0    65.0   131.0  25000.0         26.0   
7       21         9.0         6.0    65.0   179.0  27400.0         23.0   
8       16        12.0        10.0    66.0   145.0  52000.0         26.0   
...    ...         ...         ...     ...     ...      ...          ...   
12557   20        12.0        12.0    60.0   110.0   1000.0         20.0   
12565   20        12.0        11.0    63.0   152.0  38000.0         24.0   
12647   21        16.0        12.0    70.0   180.0  43600.0         30.0   
12651   21        16.0        16.0    73.0   170.0  40000.0         27.0   
12678   21         8.0         8.0    70.0   180.0  17000.0         19.0   

       FamilySize  Urban  
1               5    1.0  
5               4    1.0  
6     

2. Implementar Modelo

In [ ]:
mnb=GaussianNB() #Llamamos el modelo
mnb.fit(XD,yd)
ydp=mnb.predict(XD) #Muestra los datos de slaida - pero que el obtuvo

#Parametros del modelo por categoria
u=mnb.theta_ #Se estiman las medias
sigma=np.sqrt(mnb.var_) #Se estiman las varianzas

#Las variaciones de las variables
LI=u-sigma
LS=u+sigma

print('La media de las variables por categoria:')
print (u)
print('El limite inferior de la variación de las variables por categoria es:')
print(LI)
print('El limite superior de la variación de las variables por categoria es:')
print (LS)

#Métrica de evaluación - Como se comporto el modelo identificando rural-urban
cm=confusion_matrix(yd,ydp) #El modelo se equivoco en el 1171 falsos positivos y 4 falsos negativos
print('La matriz de confusión es:')
print(cm)


La media de las variables por categoria:
[[1.83810742e+01 1.07612958e+01 1.03324808e+01 6.72148338e+01
  1.48857630e+02 3.03672506e+04 2.20272805e+01 4.74168798e+00]
 [1.83569845e+01 1.11769401e+01 1.12565410e+01 6.69031042e+01
  1.44760976e+02 3.34278929e+04 2.26004435e+01 4.68780488e+00]]
El limite inferior de la variación de las variables por categoria es:
[[ 1.58996876e+01  7.58892579e+00  6.46311334e+00  6.31369799e+01
   1.18746964e+02 -1.33465823e+03  1.78344685e+01  2.16621834e+00]
 [ 1.58811112e+01  7.79013864e+00  7.15097738e+00  6.26897933e+01
   1.14669654e+02 -2.69533531e+03  1.84154526e+01  2.24996381e+00]]
El limite superior de la variación de las variables por categoria es:
[[2.08624608e+01 1.39336659e+01 1.42018483e+01 7.12926877e+01
  1.78968296e+02 6.20691595e+04 2.62200925e+01 7.31715762e+00]
 [2.08328577e+01 1.45637416e+01 1.53621047e+01 7.11164151e+01
  1.74852297e+02 6.95511211e+04 2.67854344e+01 7.12564594e+00]]
La matriz de confusión es:
[[   2 1171]
 [   4 450

In [ ]:
total_datos = XDB.shape[0]
print(f'El total de datos que hay en la base de datos es: {total_datos}')


El total de datos que hay en la base de datos es: 5683


In [ ]:
total_datos_urbanos = XDB[XDB['Urban'] == 1].shape[0]
print('total datos urbanos:',total_datos_urbanos)

tasa_urbanos = total_datos_urbanos / total_datos
print(f'La proporción de personas que residen en el área urbana es de: {tasa_urbanos}')

total datos urbanos: 4510
La proporción de personas que residen en el área urbana es de: 0.7935949322540912


In [ ]:
total_datos_rural = XDB[XDB['Urban'] == 0].shape[0]
print('total datos rural:',total_datos_rural)

tasa_rural = total_datos_rural / total_datos
print(f'La proporción de personas que residen en el área rural es de: {tasa_rural}')

total datos rural: 1173
La proporción de personas que residen en el área rural es de: 0.20640506774590886


**El total de datos que hay en la base de datos es:** 5683

**El total datos urbanos es:** 4510

**La proporción de personas que residen en el área urbana es de:** 0.7935949322540912

**El total datos rural es:** 1173

**La proporción de personas que residen en el área rural es de:** 0.20640506774590886


---


**Intervalo en el cual se encuentra la variacion de la edad**
*   **Limite inferior en urbano:** 1.58996876e+01
*   **Limite superior en urbano:** 2.08624608e+01
*   **Limite inferior en rural:**  1.58811112e+01
*   **Limite superior en rural:** 2.08328577e+01


**Intervalo en el cual se encuentra la altura**
*   **Limite inferior en urbano:** 6.31369799e+01
*   **Limite superior en urbano:** 7.12926877e+01
*   **Limite inferior en rural:** 6.26897933e+01
*   **Limite superior en rural:** 7.11164151e+01

**Intervalo en el cual se encuentra el peso**
*   **Limite inferior en urbano:** 1.48857630e+02
*   **Limite superior en urbano:** 1.78968296e+02
*   **Limite inferior en rural:** 1.14669654e+02
*   **Limite superior en rural:**  1.74852297e+02










3. Se procede con la evaluación del modelo

In [ ]:
VN=cm[0,0] #Verdaderos negativos
FP=cm[0,1] #Falsos positivos
FN=cm[1,0] #Falsos negativos
VP=cm[1,1] #Verdaderos positivos

Exactitud=(VP+VN)/len(XD)
print('La Exactitud es:',Exactitud) #Comportamiento general

TE=(FN+FP)/len(XD)
print('La Tasa de Error es:',TE) #Porcentaje de Equivocaciones

Sensibilidad=VP/(VP+FP)
print('La Sensibilidad es:',Sensibilidad) #Como se comporto frente a los positivos

Especificidad=VN/(VN+FN)
print('La Especificidad es:',Especificidad) #Como se comportaron frente a los negativos

La Exactitud es: 0.7932430054548654
La Tasa de Error es: 0.2067569945451346
La Sensibilidad es: 0.793729082261758
La Especificidad es: 0.3333333333333333


**La Exactitud es:** 0.7932430054548654

**La Tasa de Error es:** 0.2067569945451346

**La Sensibilidad es:** 0.793729082261758

**La Especificidad es:** 0.3333333333333333

**Análisis**

De los resultados arrojados por el modelo, se puede observar que en general, este tiene un buen desempeño, debido a que tiene una alta sensibilidad (79,37%) y exactitud (79,32%). Sin embargo, es importante mencionar que el modelo  tiene una tasa de error del 20,68%, el cual funcionó mejor clasificando
positivos, evidenciándose con la variable de la especificidad que evalúa la acertividad del modelo encontranto negativos, la cual es baja (33,33%). El modelo predijo correctamente 4508 instancias (2 verdaderos negativos + 4506 verdaderos positivos).
El modelo se equivocó en 1175 instancias (1171 falsos positivos + 4 falsos negativos).




4. ¿Qué pasa con una sola persona?

In [ ]:
Xun=np.array([[34,9,6,65,135,50000,30,3]]) #datos quiz Interactiva
ydun=mnb.predict(Xun)
print(f'Para una alguien con dichas características: {ydun}, quiere decir que vive en una zona urbana')

Para una alguien con dichas características: [1.], quiere decir que vive en una zona urbana


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(
